### ISSUE TITLE
[GJEP] `harvest` for inference debugging

### ISSUE THREAD

**Issue by femtomc** [genjax-community/genjax#237]
## Background

One way to support stateful computations in a pure language like the one JAX supports is to introduce new primitives that "raise" an effect which indicates that state is required. Then, a transformation can operate on the computation (which we can roughly think as having signature $f :: A \rightarrow \[ E \] B$ where $E$ is the effect) and dispatch with the effect by providing state.

This is the approach which GenJAX takes to implement the GFI for the `BuiltinGenerativeFunction` language.

`Oryx` also provides a transformation which is very similar to the setup I discussed above - through an interface called `harvest`. Both GenJAX's implementation of stateful handling and `harvest` rely on this transformation -- which takes pure functions with a statically raised effect to a new pure function which can accept and return out state.

Now, in GenJAX -- we access higher-order primitives (like e.g. control flow) for generative computation through generative function combinators. So our transformations for the `BuiltinGenerativeFunction` language don't attempt to cover `vmap`, or `scan`, etc. `harvest` attempts to cover these primitives - and I haven't investigated if there are bugs with the rule set, but it turns out that `harvest` could be of use in GenJAX -- more below.

## Debugging inference

One common requested operation during inference debugging is the recording/setting of state. Both @alex-lew and @georgematheos have described situations where an operation like this is useful.

The nice thing about compositional transformations on pure code is that we can actually do this, and it doesn't need to be implemented as part of the semantics of a GFI method (or a custom inference algorithm).

In this issue, I'll document some of my experiments with `harvest`, and how I think we can integrate it into a debugging interface for inference programming which is:
* Zero cost - because the fallback implementation of `sow` will be the identity function.
* Can run on any piece of code which GenJAX produces, models, inference, whatever.
* Supports nice namespace properties courtesy of Oryx's developers.

**Comment by femtomc** [comment]
Here's an example usage pattern - this just lightly wraps `harvest` to support some standard options:

<img width="908" alt="image" src="https://user-images.githubusercontent.com/34410036/219958961-fc3515de-3c41-41f7-b4da-2afd4d83996b.png">

So here, you can `tag` values to save them underneath the `grab` or `stab` transformations. `grab` will just return the values out, `stab` allows you to set and return values out.

This should seamlessly compose with other GenJAX transformations, which means we can use this inside of inference algorithms:

<img width="908" alt="image" src="https://user-images.githubusercontent.com/34410036/219959231-a52728e6-5c06-4693-a1ab-cf5bef324a20.png">

So a user can sprinkle `tag` in their code for debugging, but there's zero runtime penalty - and they can always use the `harvest` transformation later on in case they need to do some debugging.

<img width="908" alt="image" src="https://user-images.githubusercontent.com/34410036/219959669-ad95ea7b-1a87-41d2-8d08-4566c7ca7aec.png">

**Comment by femtomc** [comment]
There's a compositionality issue which I discovered with a very simple model:

```python
@genjax.gen
def model():
    x = tag(0.3, name="x")
    return Normal(x, 1.0) @ "x" # namespaces don't interact
```

Here, if you attempt to grab values out of `model.simulate` - the `harvest` transformation won't work properly because the `Tracer` values aren't handled nicely (read: compositionally) with the CPS interpreter which `model.simulate` uses (for the `Builtin` language).

One solution would be to attempt to promote the CPS interpreter to final style, and provide a principled implementation like `harvest`.

I don’t know if this is possible, because defining a continuation requires representing “the rest of the computation” — and the interfaces which JAX provides for final style tracing seem to prohibit constructing this.

Perhaps this is the reason that this actually seems to be a novel transformation (in the JAX space, I haven't found a CPS style interpreter implemented in final style).

See #239 for further details.

**Comment by femtomc** [comment]
I've discovered that the problem here is along the lines of what I thought: the CPS interpreter doesn't have its own `Trace` or `Tracer` type, so if I `bind` a primitive with a literal (like above: `tag(0.3, name="x")`) it just partially evaluates it away in the CPS transformation.

I'm hoping to figure out how to retain primitives in the CPS transformation.

**Comment by femtomc** [comment]
This comment is specific to the CPS interpreter:

Okay so the correct thing to do here is implement a trace and tracers for the continuation passing style transformation — but the interpreter _itself_ gets to raise arguments to tracers using the trace as it iterates over the Jaxpr.

then, instead of having a separate “stateful handler” data type that dispatches on primitive, we just use bind and look up rules in the context (the context acts as a stateful handler basically — so there will be a simulate context, importance context, etc).

But lifting values to the our CPS tracer will allow us to preserve other primitives that have only literals as arguments.

Furthermore, this new implementation strategy presents a nice simplification of the existing transformation — which doesn’t take full advantage of the JAX native infra for tracing. I would be more confident that the implementation doesn’t have sharp edges (e.g. for higher order primitives) if we followed the above pattern.

**Comment by femtomc** [comment]
Also: we don’t actually need CPS for the generative function interpreters — I just used a partial CPS interpreter because it closely follows the effect handling approach.

Here, for the GFI it’s likely better to just present a simple forward evaluation interpreter with the same design pattern as above — and leave the “new and improved” CPS interpreter for ADEV (for example).
